File: shared/src/geometry.ts
// Static geometry derived from the map JSON. ONE source of truth for the corner
// vertices so the client render mesh and the server/Rapier collider can never
// disagree about where a wall or ramp actually is (PRD §7, §16.1, §24).
//
// Ramp convention: `dir` is the horizontal direction the slope ASCENDS toward.
// The full-height (vertical) face is on the `dir` side; the slope descends to
// floor level on the `-dir` side. A ramp is a right-triangular prism whose two
// triangular faces are perpendicular to the other horizontal axis.

import type { Vec3, Vec3Tuple } from './math';
import { v3 } from './math';

export type RampDir = '+x' | '-x' | '+z' | '-z';

export interface SolidBox {
  type: 'box';
  pos: Vec3Tuple; // center
  size: Vec3Tuple; // full extents
}

export interface SolidRamp {
  type: 'ramp';
  pos: Vec3Tuple; // center of bounding box
  size: Vec3Tuple; // full extents of bounding box
  dir: RampDir;
}

export type Solid = SolidBox | SolidRamp;

export interface SpawnPoint {
  pos: Vec3Tuple;
  yaw: number; // degrees
}

export interface MapData {
  solids: Solid[];
  spawns: SpawnPoint[];
  killY: number;
}

/** World-space corner vertices: box -> 8, ramp -> 6 (triangular prism). */
export function solidVertices(s: Solid): Vec3[] {
  const [cx, cy, cz] = s.pos;
  const [sx, sy, sz] = s.size;
  const hx = sx / 2,
    hy = sy / 2,
    hz = sz / 2;
  const by = cy - hy;
  const ty = cy + hy;

  if (s.type === 'box') {
    return [
      v3(cx - hx, by, cz - hz),
      v3(cx + hx, by, cz - hz),
      v3(cx + hx, by, cz + hz),
      v3(cx - hx, by, cz + hz),
      v3(cx - hx, ty, cz - hz),
      v3(cx + hx, ty, cz - hz),
      v3(cx + hx, ty, cz + hz),
      v3(cx - hx, ty, cz + hz),
    ];
  }

  // Ramp: which horizontal axis the slope runs along, and which is the width.
  const alongX = s.dir === '+x' || s.dir === '-x';
  const highPos = s.dir === '+x' || s.dir === '+z'; // high face on the + side?

  // width = perpendicular horizontal axis; depth = the dir axis
  const verts: Vec3[] = [];
  for (const w of [-1, 1]) {
    if (alongX) {
      const wz = cz + w * hz; // width along z
      const toeX = highPos ? cx - hx : cx + hx; // low edge on the -dir side
      const backX = highPos ? cx + hx : cx - hx; // high edge on the +dir side
      verts.push(v3(toeX, by, wz)); // toe (low)
      verts.push(v3(backX, by, wz)); // back bottom
      verts.push(v3(backX, ty, wz)); // back top (high edge)
    } else {
      const wx = cx + w * hx; // width along x
      const toeZ = highPos ? cz - hz : cz + hz;
      const backZ = highPos ? cz + hz : cz - hz;
      verts.push(v3(wx, by, toeZ));
      verts.push(v3(wx, by, backZ));
      verts.push(v3(wx, ty, backZ));
    }
  }
  return verts;
}

/**
 * Triangle indices into solidVertices() for a closed trimesh.
 * Box -> 12 triangles (36 indices). Ramp -> 8 triangles (24 indices).
 * Winding is outward-facing CCW; render ramps DoubleSide if in doubt.
 */
export function solidTriangleIndices(s: Solid): number[] {
  if (s.type === 'box') {
    // verts: 0-3 bottom (CCW from -x-z), 4-7 top
    return [
      // bottom (y-)
      0, 2, 1, 0, 3, 2,
      // top (y+)
      4, 5, 6, 4, 6, 7,
      // -z face
      0, 1, 5, 0, 5, 4,
      // +z face
      3, 7, 6, 3, 6, 2,
      // -x face
      0, 4, 7, 0, 7, 3,
      // +x face
      1, 2, 6, 1, 6, 5,
    ];
  }
  // Ramp prism. Vertices: side A = [0:toe, 1:backBottom, 2:backTop],
  // side B = [3:toe, 4:backBottom, 5:backTop].
  return [
    // triangular faces
    0, 2, 1, // side A
    3, 4, 5, // side B
    // bottom quad (toe + backBottom)
    0, 1, 4, 0, 4, 3,
    // back vertical wall (backBottom + backTop)
    1, 2, 5, 1, 5, 4,
    // slope (toe + backTop)
    0, 3, 5, 0, 5, 2,
  ];
}

/** Flat Float32 vertex buffer + index array for building a BufferGeometry / trimesh. */
export function solidMeshArrays(s: Solid): {
  vertices: Float32Array;
  indices: Uint32Array;
} {
  const verts = solidVertices(s);
  const out = new Float32Array(verts.length * 3);
  for (let i = 0; i < verts.length; i++) {
    out[i * 3] = verts[i].x;
    out[i * 3 + 1] = verts[i].y;
    out[i * 3 + 2] = verts[i].z;
  }
  return { vertices: out, indices: new Uint32Array(solidTriangleIndices(s)) };
}

/** Walkable slope surface unit normal for a ramp (used for slide-down accel cues). */
export function rampNormal(s: SolidRamp): Vec3 {
  const [, sy] = s.size;
  const [sx, , sz] = s.size;
  // slope rises over the dir-axis run; normal points up and against the ascent.
  const alongX = s.dir === '+x' || s.dir === '-x';
  const run = alongX ? sx : sz;
  const rise = sy;
  const len = Math.hypot(run, rise) || 1;
  const ny = run / len;
  const horiz = -rise / len; // points toward the toe (down-slope horizontal)
  const sign = s.dir === '+x' || s.dir === '+z' ? 1 : -1;
  if (alongX) return v3(sign * horiz, ny, 0);
  return v3(0, ny, sign * horiz);
}

File: godot/src/Shared/Geometry.cs
// Static geometry derived from the map. ONE source of truth for the corner
// vertices so the client render mesh and the trace collider never disagree.
// Ported from shared/src/geometry.ts.
//
// Ramp convention: Dir is the horizontal direction the slope ASCENDS toward.
// The full-height (vertical) face is on the Dir side; the slope descends to floor
// level on the -Dir side. A ramp is a right-triangular prism.

using System.Collections.Generic;
using static WebRivals.Shared.VMath;

namespace WebRivals.Shared;

public enum SolidType { Box, Ramp }

public enum RampDir { PlusX, MinusX, PlusZ, MinusZ }

public sealed class Solid
{
    public SolidType Type;
    public double[] Pos;   // center [x,y,z]
    public double[] Size;  // full extents [x,y,z]
    public RampDir Dir;    // ramps only

    public static Solid Box(double px, double py, double pz, double sx, double sy, double sz)
        => new Solid { Type = SolidType.Box, Pos = new[] { px, py, pz }, Size = new[] { sx, sy, sz } };

    public static Solid Ramp(double px, double py, double pz, double sx, double sy, double sz, RampDir dir)
        => new Solid { Type = SolidType.Ramp, Pos = new[] { px, py, pz }, Size = new[] { sx, sy, sz }, Dir = dir };
}

public sealed class SpawnPoint
{
    public double[] Pos; // [x,y,z]
    public double Yaw;   // degrees
    public SpawnPoint(double x, double y, double z, double yaw) { Pos = new[] { x, y, z }; Yaw = yaw; }
}

public sealed class MapData
{
    public List<Solid> Solids = new();
    public List<SpawnPoint> Spawns = new();
    public double KillY;
}

public static class Geometry
{
    /// <summary>World-space corner vertices: box -> 8, ramp -> 6 (triangular prism).</summary>
    public static List<Vec3> SolidVertices(Solid s)
    {
        double cx = s.Pos[0], cy = s.Pos[1], cz = s.Pos[2];
        double sx = s.Size[0], sy = s.Size[1], sz = s.Size[2];
        double hx = sx / 2, hy = sy / 2, hz = sz / 2;
        double by = cy - hy;
        double ty = cy + hy;

        if (s.Type == SolidType.Box)
        {
            return new List<Vec3>
            {
                V3(cx - hx, by, cz - hz),
                V3(cx + hx, by, cz - hz),
                V3(cx + hx, by, cz + hz),
                V3(cx - hx, by, cz + hz),
                V3(cx - hx, ty, cz - hz),
                V3(cx + hx, ty, cz - hz),
                V3(cx + hx, ty, cz + hz),
                V3(cx - hx, ty, cz + hz),
            };
        }

        bool alongX = s.Dir == RampDir.PlusX || s.Dir == RampDir.MinusX;
        bool highPos = s.Dir == RampDir.PlusX || s.Dir == RampDir.PlusZ;

        var verts = new List<Vec3>();
        foreach (int w in new[] { -1, 1 })
        {
            if (alongX)
            {
                double wz = cz + w * hz;
                double toeX = highPos ? cx - hx : cx + hx;
                double backX = highPos ? cx + hx : cx - hx;
                verts.Add(V3(toeX, by, wz));
                verts.Add(V3(backX, by, wz));
                verts.Add(V3(backX, ty, wz));
            }
            else
            {
                double wx = cx + w * hx;
                double toeZ = highPos ? cz - hz : cz + hz;
                double backZ = highPos ? cz + hz : cz - hz;
                verts.Add(V3(wx, by, toeZ));
                verts.Add(V3(wx, by, backZ));
                verts.Add(V3(wx, ty, backZ));
            }
        }
        return verts;
    }

    /// <summary>Triangle indices into SolidVertices(). Box -> 12 tris, ramp -> 8.</summary>
    public static int[] SolidTriangleIndices(Solid s)
    {
        if (s.Type == SolidType.Box)
        {
            return new[]
            {
                0, 2, 1, 0, 3, 2,       // bottom
                4, 5, 6, 4, 6, 7,       // top
                0, 1, 5, 0, 5, 4,       // -z
                3, 7, 6, 3, 6, 2,       // +z
                0, 4, 7, 0, 7, 3,       // -x
                1, 2, 6, 1, 6, 5,       // +x
            };
        }
        return new[]
        {
            0, 2, 1,                    // side A
            3, 4, 5,                    // side B
            0, 1, 4, 0, 4, 3,           // bottom quad
            1, 2, 5, 1, 5, 4,           // back wall
            0, 3, 5, 0, 5, 2,           // slope
        };
    }

    /// <summary>Flat vertex buffer + index array for building a mesh / trimesh.</summary>
    public static (float[] vertices, int[] indices) SolidMeshArrays(Solid s)
    {
        var verts = SolidVertices(s);
        var outv = new float[verts.Count * 3];
        for (int i = 0; i < verts.Count; i++)
        {
            outv[i * 3] = (float)verts[i].X;
            outv[i * 3 + 1] = (float)verts[i].Y;
            outv[i * 3 + 2] = (float)verts[i].Z;
        }
        return (outv, SolidTriangleIndices(s));
    }

    /// <summary>Walkable slope surface unit normal for a ramp.</summary>
    public static Vec3 RampNormal(Solid s)
    {
        double sx = s.Size[0], sy = s.Size[1], sz = s.Size[2];
        bool alongX = s.Dir == RampDir.PlusX || s.Dir == RampDir.MinusX;
        double run = alongX ? sx : sz;
        double rise = sy;
        double len = System.Math.Sqrt(run * run + rise * rise);
        if (len == 0) len = 1;
        double ny = run / len;
        double horiz = -rise / len;
        double sign = (s.Dir == RampDir.PlusX || s.Dir == RampDir.PlusZ) ? 1 : -1;
        if (alongX) return V3(sign * horiz, ny, 0);
        return V3(0, ny, sign * horiz);
    }
}
